transform-plugins/src/main/java/co/cask/hydrator/plugin/RunExternalProgramExecutor.java

```diff
@@ -0,0 +1,256 @@
+import java.util.ArrayList;
+import java.util.List;
+import java.util.concurrent.BlockingQueue;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.Executors;
+import java.util.concurrent.LinkedBlockingQueue;
+import java.util.concurrent.TimeUnit;
+
+/**
+ * A package private class for executing an external program.
+ */
+final class RunExternalProgramExecutor extends AbstractExecutionThreadService {
+  private static final Logger LOG = LoggerFactory.getLogger(Run.class);
+  private static final long SHUTDOWN_TIMEOUT_SECONDS = 5;
+  private final String executable;
+  private final BlockingQueue<Event> eventQueue;
+  private ExecutorService executor;
+  private Process process;
+  private Thread shutdownThread;
+  private List<String> outputList = new ArrayList<>();
+
+  RunExternalProgramExecutor(String executable) {
+    this.executable = executable;
+    this.eventQueue = new LinkedBlockingQueue<Event>();
+  }
+
+  @Override
+  protected String getServiceName() {
+    return super.getServiceName();
+  }
+
+  @Override
+  protected void startUp() throws Exception {
+    // We need two threads.
+    // One thread for keep reading from input, write to process stdout and read from stdin.
+    // The other for keep reading stderr and log.
+    executor = Executors.newFixedThreadPool(2, new ThreadFactoryBuilder().setDaemon(true).build());
+
+    // The Shutdown thread is to time the shutdown and kill the process if it timeout.
+    shutdownThread = createShutdownThread();
+
+    process = Runtime.getRuntime().exec(executable);
+    executor.execute(createProcessRunnable(process));
+    executor.execute(createLogRunnable(process));
+  }
+
+  @Override
+  protected void run() throws Exception {
+    // Simply wait for the process to complete.
+    // Trigger shutdown would trigger closing of in/out streams of the process,
+    // which if the process implemented correctly, should complete.
+    int exitCode = process.waitFor();
+    if (exitCode != 0) {
+      throw new IllegalArgumentException(String.format("Process exit with exit code '%s'.", exitCode));
+    }
+  }
+
+  @Override
+  protected void triggerShutdown() {
+    executor.shutdownNow();
+    shutdownThread.start();
+  }
+
+  @Override
+  protected void shutDown() throws Exception {
+    shutdownThread.interrupt();
+    shutdownThread.join();
+
+    // Need to notify all pending events as failure.
+    List<Event> events = Lists.newLinkedList();
+    eventQueue.drainTo(events);
+    for (Event event : events) {
+      event.getCompletion().setException(
+        new IllegalStateException("External program already stopped."));
+    }
+  }
+
+  /**
+   * Sends input to the executable threads, and emits the output structured records.
+   *
+   * @param line
+   * @param completion
+   * @param emitter
+   * @param structuredRecord
+   * @param outputSchema
+   */
+  void submit(String line, SettableFuture<String> completion, Emitter<StructuredRecord> emitter,
```

Comment:
Why the future is passed it? Seems like the logic always block on one event completion (line 129)